packages/lodestar/src/chain/regen.ts

```diff
@@ -0,0 +1,205 @@
+        return latestCheckpointStateCtx;
+      }
+      return await processSlotsToNearestCheckpoint(
+        this.emitter,
+        latestCheckpointStateCtx,
+        computeStartSlotAtEpoch(this.config, cp.epoch)
+      );
+    } else {
+      // Otherwise, use the fork choice to get the stateRoot from block at the checkpoint root
+      // regenerate that state,
+      // then process empty slots until the requested epoch
+      const block = this.forkChoice.getBlock(cp.root);
+      if (!block) {
+        throw new RegenError({
+          code: RegenErrorCode.ERR_BLOCK_NOT_IN_FORKCHOICE,
+          blockRoot: cp.root,
+        });
+      }
+      const blockStateCtx = await this.getState(block.stateRoot);
+      return await processSlotsToNearestCheckpoint(
+        this.emitter,
+        blockStateCtx,
+        computeStartSlotAtEpoch(this.config, cp.epoch)
+      );
+    }
+  }
+
+  /**
+   * Return the state with `stateRoot`, minimizing the amount of regeneration needed
+   */
+  async getState(stateRoot: Root): Promise<ITreeStateContext> {
+    // Trivial case, stateCtx at stateRoot is already cached
+    const cachedStateCtx = await this.db.stateCache.get(stateRoot);
+    if (cachedStateCtx) {
+      return cachedStateCtx;
+    }
+    // Otherwise we have to use the fork choice to traverse backwards, block by block,
+    // searching the state caches
+    // then replay blocks forward to the desired stateRoot
+    const Root = this.config.types.Root;
+    const block = this.forkChoice
+      .forwardIterateBlockSummaries()
+      .find((summary) => Root.equals(summary.stateRoot, stateRoot));
+    if (!block) {
+      throw new RegenError({
+        code: RegenErrorCode.ERR_STATE_NOT_IN_FORKCHOICE,
+        stateRoot,
+      });
+    }
+    // blocks to replay, ordered highest to lowest
+    // gets reversed when replayed
+    const blocksToReplay = [block];
+    let stateCtx: ITreeStateContext | null = null;
+    for (const b of this.forkChoice.iterateBlockSummaries(block.parentRoot)) {
+      stateCtx = await this.db.stateCache.get(b.stateRoot);
+      if (stateCtx) {
+        break;
+      }
+      stateCtx = await this.db.checkpointStateCache.getLatest({
+        root: b.blockRoot,
+        epoch: computeEpochAtSlot(this.config, blocksToReplay[blocksToReplay.length - 1].slot - 1),
+      });
+      if (stateCtx) {
+        break;
+      }
+      blocksToReplay.push(b);
+    }
+    if (stateCtx === null) {
+      throw new RegenError({
+        code: RegenErrorCode.ERR_NO_SEED_STATE,
+      });
+    }
+    for (const b of blocksToReplay.reverse()) {
+      const block = await this.db.block.get(b.blockRoot);
+      if (!block) {
+        throw new RegenError({
+          code: RegenErrorCode.ERR_BLOCK_NOT_IN_DB,
+          blockRoot: b.blockRoot,
+        });
+      }
+      try {
+        stateCtx = await runStateTransition(this.emitter, this.forkChoice, stateCtx, {
```

Comment:
in case of regen, we may not want to emit anything to trigger unnecessary "onCheckpoint" at a couple of places

Reply:
I think we should emit these events any time blocks or checkpoints are processed (or reprocessed). I'd like to also trigger the `block` event here (`runStateTransition`) too.

Re: emitting events in the state regenerator, it was intentional. The thought is would help re-seed our checkpointStateCache, since that's the trigger for updating that cache. (I'd like to do the same with the `block` event and the stateCache)

If it turns out emitting events all the time isn't what we want, we have some clear paths forward. We can always hook into the job metadata to inform whether or not to emit. Eg: don't emit events when job.reprocess === true